src/SignalR/common/Shared/MessageBuffer.cs

```diff
@@ -0,0 +1,313 @@
+        {
+            // primitive backpressure if buffer is full
+            while (await _waitForAck.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
+            {
+                if (_waitForAck.Reader.TryRead(out var index)
+                    && (index == _bufferIndex || _buffer[_bufferIndex].Message is null))
+                {
+                    break;
+                }
+            }
+        }
+
+        // Avoid condition where last Ack position is the position we're currently writing into the buffer
+        // If we wrote messages around the entire buffer before another Ack arrived we would end up reading the Ack position and writing over a buffered message
+        _waitForAck.Reader.TryRead(out _);
+
+        // TODO: We could consider buffering messages until they hit backpressure in the case when the connection is down
+        await _resend.Task.ConfigureAwait(false);
+
+        var waitForResend = false;
+
+        await _writeLock.WaitAsync(cancellationToken: default).ConfigureAwait(false);
+        try
+        {
+            if (hubMessage.Message is HubInvocationMessage invocationMessage)
+            {
+                _totalMessageCount++;
+            }
+            else
+            {
+                // Non-ackable message, don't add to buffer
+                return await _connection.Transport.Output.WriteAsync(hubMessage.GetSerializedMessage(_protocol), cancellationToken).ConfigureAwait(false);
+            }
+
+            _buffer[_bufferIndex] = (hubMessage, _totalMessageCount);
+            _bufferIndex = (_bufferIndex + 1) % _buffer.Length;
+
+            return await _connection.Transport.Output.WriteAsync(hubMessage.GetSerializedMessage(_protocol), cancellationToken).ConfigureAwait(false);
+        }
+        // TODO: figure out what exception to use
+        catch (ConnectionResetException)
+        {
+            waitForResend = true;
+        }
+        finally
+        {
+            _writeLock.Release();
+        }
+
+        if (waitForResend)
+        {
+            var oldTcs = Interlocked.Exchange(ref _resend, new TaskCompletionSource<FlushResult>(TaskCreationOptions.RunContinuationsAsynchronously));
+            if (!oldTcs.Task.IsCompleted)
+            {
+                return await oldTcs.Task.ConfigureAwait(false);
+            }
+            return await _resend.Task.ConfigureAwait(false);
+        }
+
+        throw new NotImplementedException("shouldn't reach here");
+    }
+
+    public void Ack(AckMessage ackMessage)
+    {
+        // TODO: what if ackMessage.SequenceId is larger than last sent message?
+
+        // Grabbing _bufferIndex unsynchronized should be fine, we might miss the most recent message but the client shouldn't be able to ack that yet
+        // Or in exceptional cases we could miss multiple messages, but the next ack will clear them
+        var index = _bufferIndex;
+        var finalIndex = -1;
+        for (var i = 0; i < _buffer.Length; i++)
```

Comment:
This method is also only called once at a time (again enforced by calling code), it can be in parallel with other methods on `MessageBuffer`. But I think as it's currently written it is safe, unless the ValueTuple assignment to the array can have tearing since it's technically two values.

Reply:
Ok tearing is probably possible here. Will need to lock, but it shouldn't be contended too much.